File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> arr = new ArrayList<>();
        int n = genres.length;
        
        Map<String, Integer> total = new HashMap<>();
        Map<String, Map<Integer, Integer>> detail = new HashMap<>();
        
        for(int i=0;i<n;i++) {
            String genre = genres[i];
            total.put(genre, total.getOrDefault(genre, 0) + plays[i]);
            detail.computeIfAbsent(genre, k -> new HashMap()).put(i,plays[i]);
        }
        
        List<Map.Entry<String,Integer>> totalentry = new ArrayList<>(total.entrySet());
        totalentry.sort(Map.Entry.comparingByValue(Comparator.reverseOrder()));
        
        for(Map.Entry<String,Integer> t: totalentry) {
            String type = t.getKey();
            List<Map.Entry<Integer, Integer>> detailentry = new ArrayList<>(detail.get(type).entrySet());
            detailentry.sort((a,b) -> {
                if (!a.getValue().equals(b.getValue())) {
                    return b.getValue() - a.getValue();
                }
                return a.getKey() - b.getKey();
            });
            
            int cnt = 0;
            for(Map.Entry<Integer,Integer> d: detailentry) {
                arr.add(d.getKey());
                cnt ++;
                if (cnt == 2) break;
            }
        }
        
        int[] answer = new int[arr.size()];
        for(int i=0;i<arr.size();i++) {
            answer[i] = arr.get(i);
        }
        return answer;
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.swift
import Foundation

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var answer = [Int]()
    
    var total: [String: Int] = [:] // 장르별 총합
    var detail: [String: [(Int,Int)]] = [:] // plays, index
    
    for (idx, genre) in genres.enumerated() {
        total[genre, default:0] += plays[idx]
        detail[genre, default:[]].append((plays[idx],idx))
    }
    let sortedTotal = total.sorted {$0.value > $1.value}
    for x in sortedTotal {
        if var play = detail[x.key] {
            play.sort {
                if ($0.0 == $1.0) { return $0.1 < $1.1}
                return $0.0 > $1.0
            }
            if play.count >= 2 { play = Array(play[0..<2]) } 
            for (_,i) in play {
                answer.append(i)
            }
        }
    }
    
    return answer
}